R/helpers.R

```diff
@@ -698,10 +678,191 @@ h_is_positive_definite <- function(x, tolerance = 1e-08) {
+#'   function bodies are not preserved in any way after joining.
+#'
+#' @param model1 (`function`)\cr the first model to join.
+#' @param model2 (`function`)\cr the second model to join.
+#'
+#' @return joined models.
+#'
+#' @export
+#'
+h_join_models <- function(model1, model2) {
+  assert_function(model1)
+  assert_function(model2)
+  assert_class(body(model1), "{")
+  assert_class(body(model2), "{")
+
+  body2 <- as.list(body(model2))
+  if (length(body2) >= 2) {
+    body1 <- as.list(body(model1))
+    body(model1) <- as.call(c(body1, body2[-1]))
+  }
+  model1
+}
+
+#' Writing JAGS Model to a File
+#'
+#' @description `r lifecycle::badge("stable")`
+#'
+#' This function converts a R function with JAGS model into a text and then
+#' writes it into a given file. During the "model into text" conversion, the
+#' format of numbers of which absolute value is less than `0.001` or greater
+#' than `10000` is changed. These numbers will be converted into scientific
+#' format with specified number of significant digits using [`formatC`]
+#' function.
+#'
+#' @note JAGS syntax allows truncation specification like `dnorm(...) I(...)`,
+#'   which is illegal in R. To overcome this incompatibility, use dummy operator
+#'   `\%_\%` before `I(...)`, i.e. `dnorm(...) \%_\% I(...)` in the model's
+#'   code. This dummy operator `%_\%` will be removed just before saving the
+#'   JAGS code into a file.
+#'   Due to technical issues related to conversion of numbers to scientific
+#'   format, it is required that the body of a model function does not contain
+#'   `TEMP_NUM_PREF_` or `_TEMP_NUM_SUF` character constants in its body.
+#'
+#' @param model (`function`)\cr function containing the JAGS model.
+#' @param file (`string`)\cr the name of the file where the model will be saved.
+#' @param digits (`count`)\cr a desired number of significant digits for
+#'   for numbers used in JAGS input, see [`formatC`].
+#' @return Nothing, but as a side effect, the model file is written.
+#'
+#' @example examples/helpers-write_model.R
+#' @export
+#'
+h_write_model <- function(model, file = "model.jags", digits = 5) {
+  assert_function(model)
+  assert_string(file)
+  assert_count(digits)
+
+  # Replace scientific notation.
+  model_sci_replaced <- h_rapply(
+    x = body(model),
+    fun = h_format_number,
+    classes = c("integer", "numeric"),
+    digits = digits,
+    prefix = "TEMP_NUM_PREF_",
+    suffix = "_TEMP_NUM_SUF"
+  )
+  # Transform `model` body into character vector.
+  model_text <- deparse(model_sci_replaced, control = NULL)
+  model_text <- gsub("\"TEMP_NUM_PREF_|_TEMP_NUM_SUF\"", "", model_text)
+  model_text <- gsub("%_% ", "", model_text)
+  model_text <- c("model", model_text)
+
+  writeLines(model_text, con = file)
+}
+
+#' Recursively Apply a Function to a List
+#'
+#' @description `r lifecycle::badge("experimental")`
+#'
+#' This helper function is conceptually quite similar to [`base::rapply`]
```

Comment:
```suggestion
#' This helper function is conceptually quite similar to [rapply()]
```
(function is referred to with `()` and `base` does not need to be prefixed)